### PR TITLE
test: fix test_pd_client_deadlock

### PR DESCRIPTION
**Pull request by youjiali1995**
Signed-off-by: youjiali1995 <zlwgx1023@gmail.com>

### What problem does this PR solve?

Issue Number: close https://github.com/tikv/tikv/issues/9470

Problem Summary:

`region_heartbeat` in the `test_pd_client_deadlock` case sometimes hangs forever.

### What is changed and how it works?

What's Changed:

`region_heartbeat` interface in the PD client creates a stream when reconnecting to the PD leader which means `block_on`ing it can't return until reconnection. It results in the unstable test. It's changed to return a future instead of a stream which is more intuitive.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

### Release note 
* No release note

### CODE REVIEW

File: tests/failpoints/cases/test_pd_client.rs [hunk 1]
```diff
@@ -84,14 +84,14 @@ fn test_pd_client_deadlock() {
             func();
             tx.send(()).unwrap();
         });
+        // Only allow to reconnect once for a func.
```
Comment (sticnarf) [inline tests/failpoints/cases/test_pd_client.rs:87]: What's it for?

Comment (youjiali1995) [reply tests/failpoints/cases/test_pd_client.rs:87]: The reconnection interval is 100ms. It's possible the test becomes fragile due to frequent reconnection. I ban it after the first reconnection.

**Review by sticnarf — COMMENTED** [review]
/LGTM

**Review by MyonKeminta — APPROVED** [review]

**Pull request merged**